libraries/streaming-e2e-tests/react-app/src/App.js

```diff
@@ -1,12 +1,84 @@
-import React from 'react';
+import React, { Fragment, useCallback, useEffect, useMemo, useReducer } from 'react';
 import ReactWebChat from 'botframework-webchat';
+import { DirectLineStreaming } from 'botframework-directlinejs';
 
-function App({directLine}) {
-  // nonce related to botbuilder-js content security issue #2762:
-  // https://github.com/microsoft/botbuilder-js/issues/2762
-  return (
-    <ReactWebChat directLine={directLine} nonce="a1b2c3d" />
-  );
+const actions = {
+    error: 'error',
+    token: 'token',
+};
+
+function App({ bot }) {
+    if (!bot) {
+        throw new Error('bot prop is required');
+    }
+
+    const hostname = `https://${bot}.azurewebsites.net`;
+
+    const [state, dispatch] = useReducer(
```

Comment:
Not a fan of `useReducer` in this scenario because you can write it using `useState` and it's more straightforward.

```js
const [token, setToken] = useState();
const [error, setError] = useState();

// Synchronize token with hostname
useEffect(() => {
  ...
    setToken(token.token);
  } catch (err) {
    setError(`${err.message} (raw body: ${text})`);
  ...
});
```

(FYI, use `undefined` instead of `null` if possible.)

Reply:
I prefer the state manipulations to be localized in a single reducer function. For example, should fetching a new token automatically clear the error if it succeeds? Then you have to reimplement the state transformation. There's also the slight benefit of a more transactional state update - React does batch state manipulations in most cases, but [not all of them](https://jsfiddle.net/2vxm4g9z/) which can lead to unnecessary re-renders.

Reply:
This is not my code so I am fine.

Rephrasing your words, "if the React component have 2 or more `useState`, we should change all of them to a single `useReducer`". IMO, this is not exactly the intention of `useReducer`. But anyway.

Reply:
I think that's reductive - but we can agree to disagree 👍